internal/pkg/agent/application/gateway/fleet/fleet_gateway.go

```diff
@@ -494,3 +499,106 @@ func getPolicyRevisionIDX(action fleetapi.Action) int64 {
 		return 0
 	}
 }
+
+var errComponentStateChanged = errors.New("error component state changed")
+
+type StateFetcher interface {
+	// FetchState returns the current state and a context that is valid as long as the returned state is valid to use.
+	FetchState(ctx context.Context) (coordinator.State, context.Context)
+	// Done should be called once the checkin call is complete.
+	Done()
+	StartStateWatch(ctx context.Context) error
+}
+
+type FastCheckinStateFetcher struct {
+	log       *logger.Logger
+	fetcher   func() coordinator.State
+	stateChan chan coordinator.State
+
+	cancel context.CancelCauseFunc
+	mutex  sync.Mutex
+}
+
+func NewFastCheckinStateFetcher(log *logger.Logger, fetcher func() coordinator.State, stateChan chan coordinator.State) *FastCheckinStateFetcher {
+	return &FastCheckinStateFetcher{
+		log:       log,
+		fetcher:   fetcher,
+		stateChan: stateChan,
+		cancel:    nil,
+		mutex:     sync.Mutex{},
+	}
+}
+
+// Fetch wraps the state fetching to send in the check-in request under the checkin state mutex.
+// After the state is fetched the checkin cancellation function has be initialized and the new context
+// is returned.
+func (s *FastCheckinStateFetcher) FetchState(ctx context.Context) (coordinator.State, context.Context) {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	if s.cancel != nil {
+		s.cancel(nil) // ensure ctx cleanup
+	}
+
+	ctx2, ctxCancel := context.WithCancelCause(ctx)
+	state := s.fetcher()
+	s.cancel = ctxCancel
+	return state, ctx2
+}
+
+func (s *FastCheckinStateFetcher) Done() {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	if s.cancel != nil {
+		s.cancel(nil) // ensure ctx cleanup
+		s.cancel = nil
+	}
+}
+
+func (s *FastCheckinStateFetcher) invalidateState() {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	if s.cancel != nil {
+		s.cancel(errComponentStateChanged)
+		s.cancel = nil
+	}
+}
+
+func (s *FastCheckinStateFetcher) StartStateWatch(ctx context.Context) error {
+	s.log.Info("FleetGateway state watching started")
+	for {
+		select {
+		case <-ctx.Done():
+			s.log.Info("FleetGateway state watching stopped")
+			return ctx.Err()
+		case st, isOpen := <-s.stateChan:
+			if !isOpen {
+				s.log.Info("FleetGateway state watching channel closed, stopping loop.")
+				return nil
+			}
+			s.log.Infof("FleetGateway state change notification %#v", st)
```

Comment:
Logging the coordinator state is dangerous because the underlying component model contains all configuration which has in the past contained secrets.

Try to only log the parts you need. Also this has the potential to generate a lot of log spam but I do like the idea of starting to log when we start a checkin and why that was.

There is a debug log at the start of each checkin that could be tidied up to do this https://github.com/elastic/elastic-agent/blob/fcc1752e8749c34a7487f9529d8e0360a2d98fa3/internal/pkg/agent/application/gateway/fleet/fleet_gateway.go#L194

Reply:
Good catch, thanks! That logging of state was intended during development and test for verification and escaped cleanup. Will cleanup and add other logs as you suggested.

Reply:
> but I do like the idea of starting to log when we start a checkin and why that was.

We could add the a check of "last checkin outcome" and log conditionally `"Checking started"` / `"Checking started because of state change"` but since I added an explicit warining log when a check-in gets cancelled because of state change `"Check in cancelled because of state change, retrying"`, I feel it gets the job done for now. wdyt?

Reply:
@moukoublen I believe the log you already added covers it.